rfcs/text/0010_service_status.md

```diff
@@ -0,0 +1,304 @@
+in Kibana.
+
+# Basic example
+
+```ts
+// Override default behavior and only elevate severity when elasticsearch is not available
+core.status.set(
+  core.status.core$.pipe(core => core.elasticsearch);
+)
+```
+
+# Motivation
+
+Kibana should do as much possible to help users keep their installation in a working state. This includes providing as much detail about components that are not working as well as ensuring that failures in one part of the application do not block using other portions of the application.
+
+In order to provide the user with as much detail as possible about any systems that are not working correctly, the status mechanism should provide excellent defaults in terms of expressing relationships between services and presenting detailed information to the user.
+
+# Detailed design
+
+## Types
+
+```ts
+/**
+ * The current status of a service at a point in time.
+ */
+type ServiceStatus = {
+  /** The current availability level of the service. */
+  level: ServiceStatusLevel.available;
+  /** A high-level summary of the service status. */
+  summary?: string;
+  /** A more detailed description of the service status. */
+  detail?: string;
+  /** A URL to open in a new tab about how to resolve or troubleshoot the problem */
+  documentationUrl?: string;
+  /** Any JSON-serializable data to be included in the HTTP API response. */
+  meta?: object;
+} | {
+  /** The current availability level of the service. */
+  level: ServiceStatusLevel;
+  /** A high-level summary of the service status. */
+  summary: string; // required when level !== available
+  /** A more detailed description of the service status. */
+  detail?: string;
+  /** A URL to open in a new tab about how to resolve or troubleshoot the problem */
+  documentationUrl?: string;
+  /** Any JSON-serializable data to be included in the HTTP API response. */
+  meta?: object;
+}
+
+/**
+ * The current "level" of availability of a service.
+ */
+enum ServiceStatusLevel {
+  available,   // everything is working!
+  degraded,    // some features may not be working
+  unavailable, // the service is unavailable, but other functions should work
+  fatal        // block all user functions and display the status page
+}
+
+/**
+ * Status of core services.
+ */
+interface CoreStatus {
+  elasticsearch: ServiceStatus;
+  http: ServiceStatus;
+  savedObjects: ServiceStatus;
+  uiSettings: ServiceStatus;
+  metrics: ServiceStatus;
+}
+```
+
+## API Design
+
+### Plugin API
+
+```ts
+/**
+ * The API exposed to plugins on CoreSetup.status
+ */
+interface StatusSetup {
+  /**
+   * Allows a plugin to specify a custom status dependent on its own criteria.
+   * Completely overrides the default inherited status.
+   */
+  set(status$: Observable<ServiceStatus>): Observable<ServiceStatus>;
+
+  /**
+   * Current status for all Core services.
+   */
+  core$: Observable<CoreStatus>;
+
+  /**
+   * Current status for all dependencies of the current plugin.
```

Comment:
do we separate required and optional deps?

Reply:
I was thinking no, to keep it consistent with the `plugins` object received during `setup` and `start` which contains both.